---
title: "Testing close linkage v. pleiotropy"
author: "Frederick Boehm"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Setting for close linkage v pleiotropy likelihood ratio test

Our setting involves a pair of traits, $Y_1$ and $Y_2$, each of which individually (univariately) associates with a single genomic region. The exact definition of a genomic region is imprecise; in practice, it may be as large as 4 or 5 Mb. We seek to distinguish whether $Y_1$ and $Y_2$ associations (in the genomic region) arise due to a single QTL or whether there are two two distinct loci, each of which associates with exactly one of the two traits. We recognize that more complicated association patterns are possible, but we neglect them in this test.

## Installing `qtl2pleio`

We install `qtl2pleio` from github via the `devtools` R package, which is available from CRAN. 

```{r, eval = FALSE}
devtools::install_github("fboehm/qtl2pleio")
```

The above line only needs to be run once on a given computer (unless you wish to install a newer version). 

We then load the library into our R session with the `library` command:

```{r}
library(qtl2pleio)
```


We'll work with data from the `qtl2data` R package, which is on github. First, we install and load the `qtl2` package.

```{r, eval = FALSE}
devtools::install_github("rqtl/qtl2")
```

We use the above line once to install the package before loading the package with the `library` command.

```{r}
library(qtl2)
```

## Reading data from `qtl2data` R package

We read from github.com data from the `qtl2data` R package.

```{r}
tmpfile <- tempfile()
file <- paste0("https://raw.githubusercontent.com/rqtl/",
               "qtl2data/master/DOex/DOex_alleleprobs.rds")
download.file(file, tmpfile)
pr <- readRDS(tmpfile)
unlink(tmpfile)
```

We now have an allele probabilities object stored in `pr`.

```{r}
names(pr)
dim(pr$`2`)
```

We see that `pr` is a list of 3 three-dimensional arrays - one array for each of 3 chromosomes.

## Kinship calculations

For our statistical model, we need a kinship matrix. Although we don't have genome-wide data - since we have allele probabilities for only 3 chromosomes - let's calculate a kinship matrix using "leave-one-chromosome-out". In practice, one would want to use allele probabilities from a full genome-wide set of markers.

```{r}
calc_kinship(probs = pr, type = "loco") -> kinship
```

```{r}
str(kinship)
```

We see that `kinship` is a list containing 3 matrices. Each matrix is 261 by 261 - where the number of subjects is 261 - and symmetric.



## Simulating phenotypes

Before we simulate phenotype data, we first specify our statistical model.

We use the model:

$$vec(Y) = X vec(B) + vec(G) + vec(E)$$

where $Y$ is a $n$ by $2$ matrix, where each row is one subject and each column is one quantitative phenotype. $X$ is a $2n$ by $2n$ design matrix containing allele probabilities matrices for each of two (possibly identical) markers. 



## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
