<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Recla Analysis • qtl2pleio</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Recla Analysis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qtl2pleio</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/recla-analysis.html">Recla Analysis</a>
    </li>
    <li>
      <a href="../articles/testing-pleiotropy-vs-separate-qtl.html">Testing pleiotropy v separate QTL in multiparental populations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fboehm/qtl2pleio">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Recla Analysis</h1>
                        <h4 class="author">Frederick Boehm</h4>
            
            <h4 class="date">2018-11-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/fboehm/qtl2pleio/blob/master/vignettes/recla-analysis.Rmd"><code>vignettes/recla-analysis.Rmd</code></a></small>
      <div class="hidden name"><code>recla-analysis.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)</a></code></pre></div>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(qtl2pleio)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(qtl2)</a></code></pre></div>
<div id="load-recla-from-qtl2data-github-repository" class="section level2">
<h2 class="hasAnchor">
<a href="#load-recla-from-qtl2data-github-repository" class="anchor"></a>Load Recla from <code>qtl2data</code> github repository</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">file &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"https://raw.githubusercontent.com/rqtl/"</span>,</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">               <span class="st">"qtl2data/master/DO_Recla/recla.zip"</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">recla &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/read_cross2">read_cross2</a></span>(file)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co"># make sex a covariate for use in qtl2pleio::scan_pvl</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5">recla[[<span class="dv">6</span>]][ , <span class="dv">1</span>, drop =<span class="st"> </span><span class="ot">FALSE</span>] -&gt;<span class="st"> </span>sex</a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co"># insert pseudomarkers</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/insert_pseudomarkers">insert_pseudomarkers</a></span>(recla, <span class="dt">step =</span> <span class="fl">0.10</span>) -&gt;<span class="st"> </span>pseudomap</a>
<a class="sourceLine" id="cb6-8" data-line-number="8">gm &lt;-<span class="st"> </span>pseudomap<span class="op">$</span><span class="st">`</span><span class="dt">8</span><span class="st">`</span></a></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">probs &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/calc_genoprob">calc_genoprob</a></span>(recla, <span class="dt">map =</span> pseudomap)</a></code></pre></div>
<p>We now convert the genotype probabilities to haplotype dosages.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">aprobs &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/genoprob_to_alleleprob">genoprob_to_alleleprob</a></span>(probs)</a></code></pre></div>
<p>We now calculate kinship matrices, by the “leave one chromosome out (loco)” method.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">kinship &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/calc_kinship">calc_kinship</a></span>(aprobs, <span class="st">"loco"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">recla<span class="op">$</span>pheno -&gt;<span class="st"> </span>ph</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(ph) -&gt;<span class="st"> </span>lph</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(<span class="dt">FUN =</span> broman<span class="op">::</span>winsorize, <span class="dt">X =</span> lph, <span class="dt">MARGIN =</span> <span class="dv">2</span>) -&gt;<span class="st"> </span>wlph</a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="co">#colnames(wlph)[c(7, 10, 22)] &lt;- c("distance traveled in light", "percent time in light", "hot plate latency")</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="kw"><a href="http://dplyr.tidyverse.org/reference/reexports.html">as_tibble</a></span>(wlph) -&gt;<span class="st"> </span>wlph_tib</a></code></pre></div>
<p>We next perform the univariate QTL scan for our phenotypes.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">sex2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(sex <span class="op">==</span><span class="st"> "female"</span>), <span class="dt">ncol =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(sex2) &lt;-<span class="st"> "female"</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(sex2) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(aprobs[[<span class="dv">1</span>]])</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">out &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/scan1">scan1</a></span>(<span class="dt">genoprobs =</span> aprobs, <span class="dt">pheno =</span> wlph, <span class="dt">kinship =</span> kinship, <span class="dt">addcovar =</span> sex2, <span class="dt">reml =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>Let’s find the univariate QTL peaks for all phenotypes.</p>
<p>We want to look closely at those peaks on Chromosome 8. We’ll save the positions of peaks for our two traits of interest.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">(peaks &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/find_peaks">find_peaks</a></span>(out, pseudomap, <span class="dt">threshold =</span> <span class="dv">5</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(chr, pos) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="st">   </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="op">-</span><span class="st"> </span>lodindex))</a></code></pre></div>
<pre><code>##                    lodcolumn chr     pos      lod
## 1                         bw   1 23.9075 5.471580
## 2         OF_distance_first4   1 43.2385 5.772977
## 3             LD_transitions   1 95.8075 5.028258
## 4                OF_distance   2 49.9770 5.544458
## 5                         bw   2 52.3932 7.352334
## 6            OF_immobile_pct   2 53.2646 9.771813
## 7  VC_bottom_distance_first4   2 71.0160 6.531654
## 8         OF_distance_first4   3 10.7360 5.541166
## 9  VC_bottom_distance_first4   3 16.3700 5.518637
## 10           VC_top_time_pct   3 17.9390 5.951067
## 11         LD_distance_light   3 23.4390 5.203246
## 12                        bw   3 24.8390 5.632714
## 13        VC_top_time_first4   3 48.1280 6.144073
## 14           VC_top_velocity   3 48.5630 6.264824
## 15        VC_top_time_first4   4  3.5340 5.016970
## 16             OF_corner_pct   4  9.0111 6.272996
## 17           OF_immobile_pct   4 37.5206 5.529985
## 18         LD_distance_light   4 71.2992 5.040185
## 19                        bw   5 10.0740 5.728865
## 20 VC_bottom_distance_first4   5 19.9741 5.419061
## 21        VC_top_time_first4   5 20.0741 6.075282
## 22        VC_bottom_distance   5 20.5930 5.691913
## 23        VC_bottom_time_pct   5 20.5930 6.360001
## 24       TS_latency_immobile   5 43.3504 5.995115
## 25             OF_corner_pct   5 64.2551 5.665658
## 26           OF_immobile_pct   6 53.4292 6.968771
## 27     TS_frequency_climbing   6 57.0362 5.361091
## 28                        bw   7  9.1778 6.057098
## 29          TS_time_immobile   7 49.4778 8.067612
## 30        VC_bottom_distance   7 54.5591 5.011113
## 31        OF_distance_first4   7 57.9454 5.327302
## 32           VC_top_distance   7 83.8778 5.724788
## 33           OF_immobile_pct   7 83.9778 5.823556
## 34     TS_frequency_climbing   8 48.1732 5.483064
## 35         LD_distance_light   8 55.2762 5.323391
## 36              LD_light_pct   8 55.2762 5.274185
## 37                HP_latency   8 57.7732 6.223739
## 38         LD_distance_light   9 36.6965 5.196968
## 39              LD_light_pct   9 36.6965 5.417419
## 40        VC_top_time_first4   9 38.4834 5.109813
## 41           VC_top_time_pct   9 39.2680 6.356432
## 42                HP_latency   9 46.8502 5.222074
## 43                        bw  10  3.7781 6.526199
## 44        OF_distance_first4  10 29.6698 5.462975
## 45        VC_bottom_time_pct  10 32.5438 5.432804
## 46          OF_periphery_pct  10 74.8530 5.246487
## 47           VC_top_distance  11  7.8200 6.245803
## 48    VC_top_distance_first4  11 11.6236 5.486915
## 49 VC_bottom_distance_first4  11 54.3420 5.367052
## 50            LD_transitions  11 58.9000 5.903217
## 51     VC_bottom_transitions  11 60.5984 5.114051
## 52              LD_light_pct  11 63.3943 6.464176
## 53         LD_distance_light  11 63.4514 6.373437
## 54           VC_top_time_pct  12 20.5776 6.950144
## 55        VC_bottom_velocity  12 21.7760 5.653292
## 56             OF_center_pct  12 35.5140 6.399422
## 57                HP_latency  12 43.5150 5.131074
## 58          OF_periphery_pct  12 53.5776 7.240951
## 59             OF_corner_pct  13 59.7966 6.594594
## 60     VC_bottom_time_first4  14 11.9183 5.204369
## 61 VC_bottom_distance_first4  14 12.5316 5.763233
## 62     VC_bottom_transitions  14 12.5316 6.478533
## 63           VC_top_velocity  14 12.7819 6.840412
## 64        VC_bottom_distance  14 14.5316 5.592030
## 65     TS_frequency_climbing  14 21.1141 5.372594
## 66             OF_center_pct  14 53.7316 5.377182
## 67     TS_frequency_climbing  15 12.6680 6.043265
## 68              LD_light_pct  15 15.2374 5.674922
## 69        OF_distance_first4  16 23.2656 5.242101
## 70           VC_top_distance  17 15.6390 6.669275
## 71           VC_top_velocity  18  8.3750 5.558628
## 72        VC_top_time_first4  18 17.8068 6.246206
## 73            LD_transitions  18 37.4182 5.090332
## 74 VC_bottom_distance_first4  19 24.9615 7.362557
## 75     VC_bottom_time_first4  19 24.9615 7.499959
## 76           OF_immobile_pct  19 31.9505 5.639812
## 77                HP_latency  19 47.7977 5.485000</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">peaks8 &lt;-<span class="st"> </span>peaks <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(chr <span class="op">==</span><span class="st"> </span><span class="dv">8</span>, pos <span class="op">&gt;</span><span class="st"> </span><span class="dv">50</span>, pos <span class="op">&lt;</span><span class="st"> </span><span class="dv">60</span>)</a>
<a class="sourceLine" id="cb14-3" data-line-number="3">pos_LD_light_pct &lt;-<span class="st"> </span>peaks8 <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(lodcolumn <span class="op">==</span><span class="st"> "LD_light_pct"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/select.html">select</a></span>(pos)</a>
<a class="sourceLine" id="cb14-6" data-line-number="6">pos_HP_latency &lt;-<span class="st"> </span>peaks8 <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-7" data-line-number="7"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(lodcolumn <span class="op">==</span><span class="st"> "HP_latency"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-8" data-line-number="8"><span class="st">  </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/select.html">select</a></span>(pos)</a></code></pre></div>
</div>
<div id="correlation" class="section level2">
<h2 class="hasAnchor">
<a href="#correlation" class="anchor"></a>Correlation</h2>
<p>Given that the two traits “percent time in light” and “distance traveled in light” share a peak, we want to ask how correlated they are.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(wlph[ , <span class="dv">7</span>], wlph[ , <span class="dv">10</span>], <span class="dt">use =</span> <span class="st">"complete.obs"</span>)</a></code></pre></div>
<pre><code>## [1] 0.8859402</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(wlph[ , <span class="dv">22</span>], wlph[ , <span class="dv">10</span>], <span class="dt">use =</span> <span class="st">"complete.obs"</span>)</a></code></pre></div>
<pre><code>## [1] -0.1507317</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(wlph[ , <span class="dv">7</span>], wlph[ , <span class="dv">22</span>], <span class="dt">use =</span> <span class="st">"complete.obs"</span>)</a></code></pre></div>
<pre><code>## [1] -0.143598</code></pre>
</div>
<div id="scatter-plot-of-phenotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#scatter-plot-of-phenotypes" class="anchor"></a>Scatter plot of phenotypes</h2>
<p>We create a scatter plot for the two phenotypes.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">scatter1022 &lt;-<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">data =</span> wlph_tib, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">y =</span> HP_latency, <span class="dt">x =</span> LD_light_pct)) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="dt">x =</span> <span class="st">"Percent time in light"</span>, <span class="dt">y =</span> <span class="st">"Hot plate latency"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Scatterplot of hot plate latency vs. percent time in light"</span>)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">scatter1022</a></code></pre></div>
<pre><code>## Warning: Removed 3 rows containing missing values (geom_point).</code></pre>
<p><img src="recla-analysis_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div id="genome-wide-lod-plots-for-the-traits-from-recla" class="section level2">
<h2 class="hasAnchor">
<a href="#genome-wide-lod-plots-for-the-traits-from-recla" class="anchor"></a>Genome-wide LOD plots for the traits from Recla</h2>
<p>Let’s plot the results of the univariate QTL scans for our two traits.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(out, <span class="dt">map =</span> pseudomap, <span class="dt">lodcolumn =</span> <span class="dv">10</span>, <span class="dt">main =</span> <span class="st">"Genome-wide LOD for percent time in light"</span>)</a></code></pre></div>
<p><img src="recla-analysis_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(out, <span class="dt">map =</span> pseudomap, <span class="dt">lodcolumn =</span> <span class="dv">22</span>, <span class="dt">main =</span> <span class="st">"Genome-wide LOD for hot plate latency"</span>)</a></code></pre></div>
<p><img src="recla-analysis_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div id="allele-effects-plots-on-chr-8-for-each-of-the-three-recla-traits" class="section level2">
<h2 class="hasAnchor">
<a href="#allele-effects-plots-on-chr-8-for-each-of-the-three-recla-traits" class="anchor"></a>Allele effects plots on chr 8 for each of the three Recla traits</h2>
<p>We examine the allele effects plots for our two traits, in the region of interest on Chromosome 8.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/scan1coef">scan1coef</a></span>(aprobs[ , <span class="dv">8</span>], <span class="dt">pheno =</span> wlph[, <span class="dv">10</span>], <span class="dt">kinship =</span> kinship<span class="op">$</span><span class="st">`</span><span class="dt">8</span><span class="st">`</span>, </a>
<a class="sourceLine" id="cb25-2" data-line-number="2">          <span class="dt">reml =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb25-3" data-line-number="3">          <span class="dt">addcovar =</span> sex2) -&gt;<span class="st"> </span>s1c_<span class="dv">10</span></a>
<a class="sourceLine" id="cb25-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/scan1coef">scan1coef</a></span>(aprobs[ , <span class="dv">8</span>], <span class="dt">pheno =</span> wlph[, <span class="dv">22</span>], <span class="dt">kinship =</span> kinship<span class="op">$</span><span class="st">`</span><span class="dt">8</span><span class="st">`</span>, </a>
<a class="sourceLine" id="cb25-5" data-line-number="5">          <span class="dt">reml =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb25-6" data-line-number="6">          <span class="dt">addcovar =</span> sex2) -&gt;<span class="st"> </span>s1c_<span class="dv">22</span></a></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="co"># subset scan1output objects</span></a>
<a class="sourceLine" id="cb26-2" data-line-number="2">s1c_10s &lt;-<span class="st"> </span>s1c_<span class="dv">10</span>[<span class="dv">650</span><span class="op">:</span><span class="dv">999</span>, ] <span class="co"># 650:999 is the same as the interval for the two-dimensional scan.</span></a>
<a class="sourceLine" id="cb26-3" data-line-number="3">s1c_22s &lt;-<span class="st"> </span>s1c_<span class="dv">22</span>[<span class="dv">650</span><span class="op">:</span><span class="dv">999</span>, ]</a></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/plot_coef">plot_coefCC</a></span>(s1c_10s, <span class="dt">map =</span> pseudomap, <span class="dt">main =</span> <span class="st">"Allele effects for percent time in light"</span>, <span class="dt">scan1_output =</span> s1c_10s)</a></code></pre></div>
<p><img src="recla-analysis_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/qtl2/topics/plot_coef">plot_coefCC</a></span>(s1c_22s, <span class="dt">map =</span> pseudomap, <span class="dt">main =</span> <span class="st">"Allele effects for hot plate latency"</span>, <span class="dt">scan1_output =</span> s1c_22s)</a></code></pre></div>
<p><img src="recla-analysis_files/figure-html/unnamed-chunk-15-2.png" width="700"></p>
</div>
<div id="two-dimensional-scan-results-from-github" class="section level2">
<h2 class="hasAnchor">
<a href="#two-dimensional-scan-results-from-github" class="anchor"></a>Two-dimensional scan results from github</h2>
<p>We present the code that we ran to perform the two-dimensional scan.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw"><a href="../reference/scan_pvl.html">scan_pvl</a></span>(<span class="dt">probs =</span> pp, <span class="dt">pheno =</span> wlph[, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">10</span>, <span class="dv">22</span>)], <span class="dt">covariates =</span> sex2, <span class="dt">kinship =</span> kinship<span class="op">$</span><span class="st">`</span><span class="dt">8</span><span class="st">`</span>, <span class="dt">start_snp1 =</span> <span class="dv">650</span>, <span class="dt">n_snp =</span> <span class="dv">350</span>) -&gt;<span class="st"> </span>pvl1022</a>
<a class="sourceLine" id="cb29-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(<span class="dt">x =</span> pvl1022, <span class="dt">file =</span> <span class="st">"recla-10-22.txt"</span>)</a></code></pre></div>
<p>To save computing time, we read the two-dimensional scan results from github.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="kw"><a href="http://dplyr.tidyverse.org/reference/reexports.html">as_tibble</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"https://raw.githubusercontent.com/fboehm/qtl2pleio-manuscript/master/Rmd/recla-10-22.txt"</span>)) -&gt;<span class="st"> </span>pvl1022</a></code></pre></div>
<p>We then calculate the likelihood ratio test statistic.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1">(mylrt &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calc_lrt_tib.html">calc_lrt_tib</a></span>(pvl1022))</a></code></pre></div>
<pre><code>## [1] 2.771408</code></pre>
</div>
<div id="profile-lod-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#profile-lod-plot" class="anchor"></a>Profile LOD Plot</h2>
<p>We create a profile LOD plot.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(recla<span class="op">$</span>pheno)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">10</span>, <span class="dv">22</span>)] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Percent time in light"</span>, <span class="st">"Hot plate latency"</span>)</a>
<a class="sourceLine" id="cb33-2" data-line-number="2">p1022 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tidy_scan_pvl.html">tidy_scan_pvl</a></span>(pvl1022, <span class="dt">pmap =</span> gm) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb33-3" data-line-number="3"><span class="st">  </span><span class="kw"><a href="../reference/add_intercepts.html">add_intercepts</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(pos_LD_light_pct), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(pos_HP_latency))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb33-4" data-line-number="4"><span class="st">  </span><span class="kw"><a href="../reference/plot_pvl.html">plot_pvl</a></span>(<span class="dt">phenames =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(recla<span class="op">$</span>pheno)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">10</span>, <span class="dv">22</span>)]) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Chromosome 8 profile LOD for percent time in light and hot plate latency"</span>) </a></code></pre></div>
<pre><code>## Warning: Column `marker1`/`marker` joining factor and character vector,
## coercing into character vector</code></pre>
<pre><code>## Warning: Column `marker2`/`marker` joining factor and character vector,
## coercing into character vector</code></pre>
</div>
<div id="bootstrap-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#bootstrap-analyses" class="anchor"></a>Bootstrap analyses</h2>
<p>First, we find the <em>pleiotropy peak</em> marker. This is the marker for which the log likelihood is maximized under the constraint of pleiotropy.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw"><a href="../reference/find_pleio_peak_tib.html">find_pleio_peak_tib</a></span>(pvl1022, <span class="dt">start_snp =</span> <span class="dv">650</span>)</a></code></pre></div>
<pre><code>## loglik139 
##       788</code></pre>
<p>To save computing time, we read the bootstrap results files from Github. For details of how we performed the bootstrap analyses on the University of Wisconsin-Madison Center for High-Throughput Computing, please see the documentation in the qtl2pleio-manuscript repository: <a href="https://github.com/fboehm/qtl2pleio-manuscript" class="uri">https://github.com/fboehm/qtl2pleio-manuscript</a>.</p>
<p>The code below creates a temporary directory “tmp” in the user’s working directory. We then download a gzipped tar file that contains 1000 text files. Each text file contains a single likelihood ratio test statistic from a bootstrap sample.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1">gz_file &lt;-<span class="st"> "https://raw.githubusercontent.com/fboehm/qtl2pleio-manuscript/master/chtc/Recla-bootstrap/submit_files/recla-boot-run561.tar.gz"</span></a>
<a class="sourceLine" id="cb38-2" data-line-number="2">tmp_dir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"tmp"</span>)</a>
<a class="sourceLine" id="cb38-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(tmp_dir)</a>
<a class="sourceLine" id="cb38-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/download.file">download.file</a></span>(gz_file, <span class="dt">destfile =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(tmp_dir, <span class="st">"recla-boot-run561.tar.gz"</span>))</a>
<a class="sourceLine" id="cb38-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/untar">untar</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(tmp_dir, <span class="st">"recla-boot-run561.tar.gz"</span>), <span class="dt">exdir =</span> tmp_dir)</a>
<a class="sourceLine" id="cb38-6" data-line-number="6"><span class="co">## read boot lrt files</span></a>
<a class="sourceLine" id="cb38-7" data-line-number="7">boot_lrt &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>()</a>
<a class="sourceLine" id="cb38-8" data-line-number="8"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>){</a>
<a class="sourceLine" id="cb38-9" data-line-number="9">  n &lt;-<span class="st"> </span>i <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb38-10" data-line-number="10">  fn &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"recla-boot-run561_"</span>, n, <span class="st">".txt"</span>)</a>
<a class="sourceLine" id="cb38-11" data-line-number="11">  boot_lrt[i] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(tmp_dir, fn))</a>
<a class="sourceLine" id="cb38-12" data-line-number="12">}</a>
<a class="sourceLine" id="cb38-13" data-line-number="13"><span class="co"># convert list to numeric vector</span></a>
<a class="sourceLine" id="cb38-14" data-line-number="14">boot_lrt &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(boot_lrt)</a>
<a class="sourceLine" id="cb38-15" data-line-number="15"><span class="co"># delete tmp_dir and its contents</span></a>
<a class="sourceLine" id="cb38-16" data-line-number="16"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlink">unlink</a></span>(tmp_dir, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p>We get a bootstrap p-value by comparing the above vector’s values to <code>mylrt</code>, the test statistic for the observed data.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>(boot_lrt <span class="op">&gt;=</span><span class="st"> </span>mylrt) <span class="op">/</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(boot_lrt)</a></code></pre></div>
<pre><code>## [1] 0.109</code></pre>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">session_info</a></span>()</a></code></pre></div>
<pre><code>## ─ Session info ──────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 3.5.1 (2018-07-02)
##  os       macOS  10.14.1              
##  system   x86_64, darwin15.6.0        
##  ui       X11                         
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  ctype    en_US.UTF-8                 
##  tz       America/Chicago             
##  date     2018-11-21                  
## 
## ─ Packages ──────────────────────────────────────────────────────────────
##  package     * version    date       lib source                         
##  assertthat    0.2.0      2017-04-11 [1] CRAN (R 3.5.0)                 
##  backports     1.1.2      2017-12-13 [1] CRAN (R 3.5.0)                 
##  base64enc     0.1-3      2015-07-28 [1] CRAN (R 3.5.0)                 
##  bindr         0.1.1      2018-03-13 [1] CRAN (R 3.5.0)                 
##  bindrcpp    * 0.2.2      2018-03-29 [1] CRAN (R 3.5.0)                 
##  bit           1.1-14     2018-05-29 [1] CRAN (R 3.5.0)                 
##  bit64         0.9-7      2017-05-08 [1] CRAN (R 3.5.0)                 
##  blob          1.1.1      2018-03-25 [1] CRAN (R 3.5.0)                 
##  broman        0.68-2     2018-07-25 [1] CRAN (R 3.5.0)                 
##  callr         3.0.0      2018-08-24 [1] CRAN (R 3.5.0)                 
##  cli           1.0.1      2018-09-25 [1] CRAN (R 3.5.0)                 
##  colorspace    1.3-2      2016-12-14 [1] CRAN (R 3.5.0)                 
##  commonmark    1.6        2018-09-30 [1] CRAN (R 3.5.0)                 
##  crayon        1.3.4      2017-09-16 [1] CRAN (R 3.5.0)                 
##  data.table    1.11.8     2018-09-30 [1] CRAN (R 3.5.0)                 
##  DBI           1.0.0      2018-05-02 [1] CRAN (R 3.5.0)                 
##  desc          1.2.0      2018-05-01 [1] CRAN (R 3.5.0)                 
##  devtools      2.0.1      2018-10-26 [1] CRAN (R 3.5.1)                 
##  digest        0.6.18     2018-10-10 [1] CRAN (R 3.5.0)                 
##  dplyr       * 0.7.8      2018-11-10 [1] CRAN (R 3.5.0)                 
##  evaluate      0.12       2018-10-09 [1] CRAN (R 3.5.0)                 
##  fs            1.2.6      2018-08-23 [1] CRAN (R 3.5.0)                 
##  ggplot2     * 3.1.0      2018-10-25 [1] CRAN (R 3.5.0)                 
##  glue          1.3.0      2018-07-17 [1] CRAN (R 3.5.0)                 
##  gtable        0.2.0      2016-02-26 [1] CRAN (R 3.5.0)                 
##  htmltools     0.3.6      2017-04-28 [1] CRAN (R 3.5.0)                 
##  jsonlite      1.5        2017-06-01 [1] CRAN (R 3.5.0)                 
##  knitr         1.20       2018-02-20 [1] CRAN (R 3.5.0)                 
##  labeling      0.3        2014-08-23 [1] CRAN (R 3.5.0)                 
##  lazyeval      0.2.1      2017-10-29 [1] CRAN (R 3.5.0)                 
##  magrittr      1.5        2014-11-22 [1] CRAN (R 3.5.0)                 
##  MASS          7.3-50     2018-04-30 [1] CRAN (R 3.5.1)                 
##  memoise       1.1.0      2017-04-21 [1] CRAN (R 3.5.0)                 
##  munsell       0.5.0      2018-06-12 [1] CRAN (R 3.5.0)                 
##  pillar        1.3.0      2018-07-14 [1] CRAN (R 3.5.0)                 
##  pkgbuild      1.0.2      2018-10-16 [1] CRAN (R 3.5.0)                 
##  pkgconfig     2.0.2      2018-08-16 [1] CRAN (R 3.5.0)                 
##  pkgdown       1.2.0.9000 2018-11-21 [1] Github (hadley/pkgdown@44460e6)
##  pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.5.0)                 
##  plyr          1.8.4      2016-06-08 [1] CRAN (R 3.5.0)                 
##  prettyunits   1.0.2      2015-07-13 [1] CRAN (R 3.5.0)                 
##  processx      3.2.0      2018-08-16 [1] CRAN (R 3.5.0)                 
##  ps            1.2.1      2018-11-06 [1] CRAN (R 3.5.0)                 
##  purrr         0.2.5      2018-05-29 [1] CRAN (R 3.5.0)                 
##  qtl2        * 0.17-9     2018-11-18 [1] Github (rqtl/qtl2@1c007a2)     
##  qtl2pleio   * 0.1.2.9000 2018-11-21 [1] local                          
##  R6            2.3.0      2018-10-04 [1] CRAN (R 3.5.0)                 
##  Rcpp          1.0.0.1    2018-11-18 [1] Github (RcppCore/Rcpp@4f168e6) 
##  remotes       2.0.2      2018-10-30 [1] CRAN (R 3.5.0)                 
##  rlang         0.3.0.1    2018-10-25 [1] CRAN (R 3.5.0)                 
##  rmarkdown     1.10       2018-06-11 [1] CRAN (R 3.5.0)                 
##  roxygen2      6.1.1      2018-11-07 [1] CRAN (R 3.5.0)                 
##  rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.5.0)                 
##  RSQLite       2.1.1      2018-05-06 [1] CRAN (R 3.5.0)                 
##  rstudioapi    0.8        2018-10-02 [1] CRAN (R 3.5.0)                 
##  scales        1.0.0      2018-08-09 [1] CRAN (R 3.5.0)                 
##  sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 3.5.0)                 
##  stringi       1.2.4      2018-07-20 [1] CRAN (R 3.5.0)                 
##  stringr       1.3.1      2018-05-10 [1] CRAN (R 3.5.0)                 
##  testthat      2.0.1      2018-10-13 [1] CRAN (R 3.5.0)                 
##  tibble        1.4.2      2018-01-22 [1] CRAN (R 3.5.0)                 
##  tidyselect    0.2.5      2018-10-11 [1] CRAN (R 3.5.0)                 
##  usethis       1.4.0      2018-08-14 [1] CRAN (R 3.5.0)                 
##  withr         2.1.2      2018-03-15 [1] CRAN (R 3.5.0)                 
##  xml2          1.2.0      2018-01-24 [1] CRAN (R 3.5.0)                 
##  yaml          2.2.0      2018-07-25 [1] CRAN (R 3.5.0)                 
## 
## [1] /Library/Frameworks/R.framework/Versions/3.5/Resources/library</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-recla-from-qtl2data-github-repository">Load Recla from <code>qtl2data</code> github repository</a></li>
      <li><a href="#correlation">Correlation</a></li>
      <li><a href="#scatter-plot-of-phenotypes">Scatter plot of phenotypes</a></li>
      <li><a href="#genome-wide-lod-plots-for-the-traits-from-recla">Genome-wide LOD plots for the traits from Recla</a></li>
      <li><a href="#allele-effects-plots-on-chr-8-for-each-of-the-three-recla-traits">Allele effects plots on chr 8 for each of the three Recla traits</a></li>
      <li><a href="#two-dimensional-scan-results-from-github">Two-dimensional scan results from github</a></li>
      <li><a href="#profile-lod-plot">Profile LOD Plot</a></li>
      <li><a href="#bootstrap-analyses">Bootstrap analyses</a></li>
      <li><a href="#session-info">Session info</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Frederick Boehm.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
