<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Testing close linkage v. pleiotropy • qtl2pleio</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Testing close linkage v. pleiotropy">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qtl2pleio</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/testing-close-linkage-vs-pleiotropy.html">Testing close linkage v. pleiotropy</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Testing close linkage v. pleiotropy</h1>
                        <h4 class="author">Frederick Boehm</h4>
            
            <h4 class="date">2018-04-04</h4>
      
      
      <p class="hidden name"><code>testing-close-linkage-vs-pleiotropy.Rmd</code></p>

    </div>

    
    
<div id="setting-for-close-linkage-v-pleiotropy-likelihood-ratio-test" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-for-close-linkage-v-pleiotropy-likelihood-ratio-test" class="anchor"></a>Setting for close linkage v pleiotropy likelihood ratio test</h2>
<p>Our setting involves a pair of traits, <span class="math inline">\(Y_1\)</span> and <span class="math inline">\(Y_2\)</span>, each of which individually (univariately) maps to a single genomic region. <span class="math inline">\(Y_1\)</span> and <span class="math inline">\(Y_2\)</span> are both measured on the same subjects. The exact definition of a genomic region is imprecise; in practice, it may be as large as 4 or 5 Mb. We seek to distinguish whether <span class="math inline">\(Y_1\)</span> and <span class="math inline">\(Y_2\)</span> associations (in the genomic region of interest) arise due to a single QTL or whether there are two two distinct loci, each of which associates with exactly one of the two traits. We recognize that more complicated association patterns are possible, but we neglect them in this test.</p>
</div>
<div id="installing-qtl2pleio" class="section level2">
<h2 class="hasAnchor">
<a href="#installing-qtl2pleio" class="anchor"></a>Installing <code>qtl2pleio</code>
</h2>
<p>We install <code>qtl2pleio</code> from github via the <code>devtools</code> R package, which is available from CRAN.</p>
<p>If you haven’t installed <code>devtools</code> R package, you can do so with this line of code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)</code></pre></div>
<p>Now that you have installed <code>devtools</code>, you can install <code>qtl2pleio</code> from its Github repository with this line of code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"fboehm/qtl2pleio"</span>)</code></pre></div>
<p>The above line only needs to be run once on a given computer (unless you wish to install a newer version of the package).</p>
<p>We then load the library into our R session with the <code>library</code> command:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(qtl2pleio)</code></pre></div>
<p>We’ll work with data from the <code>qtl2data</code> R package, which is on github. First, we install and load the <code>qtl2</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"rqtl/qtl2"</span>)</code></pre></div>
<p>We use the above line once to install the package on our computer before loading the package with the <code>library</code> command.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(qtl2)</code></pre></div>
</div>
<div id="reading-data-from-qtl2data-r-package" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-data-from-qtl2data-r-package" class="anchor"></a>Reading data from <code>qtl2data</code> R package</h2>
<p>We read from github.com data from the <code>qtl2data</code> R package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tmpfile &lt;-<span class="st"> </span><span class="kw">tempfile</span>()
file &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">"https://raw.githubusercontent.com/rqtl/"</span>,
               <span class="st">"qtl2data/master/DOex/DOex_alleleprobs.rds"</span>)
<span class="kw">download.file</span>(file, tmpfile)
pr &lt;-<span class="st"> </span><span class="kw">readRDS</span>(tmpfile)
<span class="kw">unlink</span>(tmpfile)
tmpfile &lt;-<span class="st"> </span><span class="kw">tempfile</span>()
file &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">"https://raw.githubusercontent.com/rqtl/"</span>,
               <span class="st">"qtl2data/master/DOex/DOex_pmap.csv"</span>)
<span class="kw">download.file</span>(file, tmpfile)
pmap_pre &lt;-<span class="st"> </span><span class="kw">read.csv</span>(tmpfile)
<span class="kw">unlink</span>(tmpfile)
pm2 &lt;-<span class="st"> </span>pmap_pre[pmap_pre<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span><span class="dv">2</span>, <span class="dv">3</span>]
<span class="kw">names</span>(pm2) &lt;-<span class="st"> </span>pmap_pre[pmap_pre<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span><span class="dv">2</span>, <span class="dv">1</span>]
pm3 &lt;-<span class="st"> </span>pmap_pre[pmap_pre<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span><span class="dv">3</span>, <span class="dv">3</span>]
<span class="kw">names</span>(pm3) &lt;-<span class="st"> </span>pmap_pre[pmap_pre<span class="op">$</span>chr <span class="op">==</span><span class="st"> </span><span class="dv">3</span>, <span class="dv">1</span>]
pmX &lt;-<span class="st"> </span>pmap_pre[pmap_pre<span class="op">$</span>chr <span class="op">==</span><span class="st"> "X"</span>, <span class="dv">3</span>]
<span class="kw">names</span>(pmX) &lt;-<span class="st"> </span>pmap_pre[pmap_pre<span class="op">$</span>chr <span class="op">==</span><span class="st"> "X"</span>, <span class="dv">1</span>]
<span class="kw">list</span>(pm2, pm3, pmX) -&gt;<span class="st"> </span>pm
<span class="kw">names</span>(pm) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"`2`"</span>, <span class="st">"`3`"</span>, <span class="st">"X"</span>)</code></pre></div>
<p>We now have an allele probabilities object stored in <code>pr</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(pr)
<span class="co">#&gt; [1] "2" "3" "X"</span>
<span class="kw">dim</span>(pr<span class="op">$</span><span class="st">`</span><span class="dt">2</span><span class="st">`</span>)
<span class="co">#&gt; [1] 261   8 125</span></code></pre></div>
<p>We see that <code>pr</code> is a list of 3 three-dimensional arrays - one array for each of 3 chromosomes.</p>
</div>
<div id="kinship-calculations" class="section level2">
<h2 class="hasAnchor">
<a href="#kinship-calculations" class="anchor"></a>Kinship calculations</h2>
<p>For our statistical model, we need a kinship matrix. Although we don’t have genome-wide data - since we have allele probabilities for only 3 chromosomes - let’s calculate a kinship matrix using “leave-one-chromosome-out”. In practice, one would want to use allele probabilities from a full genome-wide set of markers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/qtl2/topics/calc_kinship">calc_kinship</a></span>(<span class="dt">probs =</span> pr, <span class="dt">type =</span> <span class="st">"loco"</span>) -&gt;<span class="st"> </span>kinship</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(kinship)
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ 2: num [1:261, 1:261] 0.6889 0.0734 0.2169 0.0517 0.0488 ...</span>
<span class="co">#&gt;   ..- attr(*, "n_pos")= int 195</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:261] "1" "4" "5" "6" ...</span>
<span class="co">#&gt;   .. ..$ : chr [1:261] "1" "4" "5" "6" ...</span>
<span class="co">#&gt;  $ 3: num [1:261, 1:261] 0.654 0.0732 0.2084 0.0917 0.0772 ...</span>
<span class="co">#&gt;   ..- attr(*, "n_pos")= int 218</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:261] "1" "4" "5" "6" ...</span>
<span class="co">#&gt;   .. ..$ : chr [1:261] "1" "4" "5" "6" ...</span>
<span class="co">#&gt;  $ X: num [1:261, 1:261] 0.4694 0.0714 0.2058 0.098 0.1135 ...</span>
<span class="co">#&gt;   ..- attr(*, "n_pos")= int 227</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:261] "1" "4" "5" "6" ...</span>
<span class="co">#&gt;   .. ..$ : chr [1:261] "1" "4" "5" "6" ...</span></code></pre></div>
<p>We see that <code>kinship</code> is a list containing 3 matrices. Each matrix is 261 by 261 - where the number of subjects is 261 - and symmetric. The <span class="math inline">\((i, j)\)</span> cell in the matrix contains the estimate of identity-by-state (IBS) probability for randomly chosen alleles at a given locus for those two subjects.</p>
<p>Before we simulate phenotype data, we first specify our statistical model.</p>
<p>We use the model:</p>
<p><span class="math display">\[vec(Y) = X vec(B) + vec(G) + vec(E)\]</span></p>
<p>where <span class="math inline">\(Y\)</span> is a <span class="math inline">\(n\)</span> by <span class="math inline">\(2\)</span> matrix, where each row is one subject and each column is one quantitative phenotype. <span class="math inline">\(X\)</span> is a <span class="math inline">\(2n\)</span> by <span class="math inline">\(2f\)</span> design matrix containing <span class="math inline">\(n\)</span> by <span class="math inline">\(f\)</span> allele probabilities matrices for each of two (possibly identical) markers. Thus, <span class="math inline">\(X\)</span> is a block-diagonal matrix, with exactly two <span class="math inline">\(n\)</span> by <span class="math inline">\(f\)</span> blocks on the diagonal. <span class="math inline">\(B\)</span> is <span class="math inline">\(f\)</span> by 2 matrix. “vec” refers to the vectorization operator. “vec(B)”, where <span class="math inline">\(B\)</span> is a <span class="math inline">\(f\)</span> by <span class="math inline">\(2\)</span> matrix, is, thus, a (column) vector of length <span class="math inline">\(2f\)</span> that is formed by stacking the second column of <span class="math inline">\(B\)</span> beneath the first column of <span class="math inline">\(B\)</span>.</p>
<p><span class="math inline">\(G\)</span> is a matrix of random effects. We specify its distribution as matrix-variate normal with mean being a <span class="math inline">\(n\)</span> by <span class="math inline">\(2\)</span> matrix of zeros, covariance among row vectors a <span class="math inline">\(n\)</span> by <span class="math inline">\(n\)</span> kinship matrix, <span class="math inline">\(K\)</span>, and covariance among column vectors a <span class="math inline">\(2\)</span> by <span class="math inline">\(2\)</span> genetic covariance matrix, <span class="math inline">\(V_g\)</span>.</p>
<p>In mathematical notation, we write:</p>
<p><span class="math display">\[G \sim MN_{\text{n by 2}}(0, K, V_g)\]</span></p>
<p>We also need to specify the distribution of the <span class="math inline">\(E\)</span> matrix, which contains the random errors. <span class="math inline">\(E\)</span> is a random <span class="math inline">\(n\)</span> by <span class="math inline">\(2\)</span> matrix that is distributed as a matrix-variate normal distribution with mean being the <span class="math inline">\(n\)</span> by <span class="math inline">\(2\)</span> zero matrix, covariance among row vectors <span class="math inline">\(I_n\)</span>, the <span class="math inline">\(n\)</span> by <span class="math inline">\(n\)</span> identity matrix, and covariance among columns the <span class="math inline">\(2\)</span> by <span class="math inline">\(2\)</span> matrix <span class="math inline">\(V_e\)</span>.</p>
<p>In practice, we typically observe the phenotype matrix <span class="math inline">\(Y\)</span>. We also treat as known the design matrix <span class="math inline">\(X\)</span> and the kinship matrix <span class="math inline">\(K\)</span>. We then infer the values of <span class="math inline">\(B\)</span>, <span class="math inline">\(V_g\)</span>, and <span class="math inline">\(V_e\)</span>.</p>
</div>
<div id="simulating-phenotypes-with-qtl2pleiosim1" class="section level2">
<h2 class="hasAnchor">
<a href="#simulating-phenotypes-with-qtl2pleiosim1" class="anchor"></a>Simulating phenotypes with <code><a href="../reference/sim1.html">qtl2pleio::sim1</a></code>
</h2>
<p>The function to simulate phenotypes in <code>qtl2pleio</code> is <code>sim1</code>. By examining its help page, we see that it takes five arguments. The help page also gives the dimensions of the inputs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># set up the design matrix, X</span>
pp &lt;-<span class="st"> </span>pr<span class="op">$</span><span class="st">`</span><span class="dt">2</span><span class="st">`</span>
X &lt;-<span class="st"> </span>gemma2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/gemma2/topics/stagger_mats">stagger_mats</a></span>(pp[ , , <span class="dv">50</span>], pp[ , , <span class="dv">50</span>])
<span class="co"># assemble B matrix of allele effects</span>
B &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dt">data =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>), <span class="dt">nrow =</span> <span class="dv">8</span>, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">byrow =</span> <span class="ot">FALSE</span>)
<span class="co"># verify that B is what we want:</span>
B
<span class="co">#&gt;      [,1] [,2]</span>
<span class="co">#&gt; [1,]   -1   -1</span>
<span class="co">#&gt; [2,]   -1   -1</span>
<span class="co">#&gt; [3,]   -1   -1</span>
<span class="co">#&gt; [4,]   -1   -1</span>
<span class="co">#&gt; [5,]    1    1</span>
<span class="co">#&gt; [6,]    1    1</span>
<span class="co">#&gt; [7,]    1    1</span>
<span class="co">#&gt; [8,]    1    1</span>
<span class="co"># set.seed to ensure reproducibility</span>
<span class="kw">set.seed</span>(<span class="dv">2018</span><span class="op">-</span><span class="dv">01</span><span class="op">-</span><span class="dv">30</span>)
<span class="co"># call to sim1</span>
Ypre &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sim1.html">sim1</a></span>(<span class="dt">X =</span> X, <span class="dt">B =</span> B, <span class="dt">Vg =</span> <span class="kw">diag</span>(<span class="dv">2</span>), <span class="dt">Ve =</span> <span class="kw">diag</span>(<span class="dv">2</span>), <span class="dt">kinship =</span> kinship[[<span class="dv">1</span>]])
Y &lt;-<span class="st"> </span><span class="kw">matrix</span>(Ypre, <span class="dt">nrow =</span> <span class="dv">261</span>, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">byrow =</span> <span class="ot">FALSE</span>)
<span class="kw">rownames</span>(Y) &lt;-<span class="st"> </span><span class="kw">rownames</span>(pp)
<span class="kw">colnames</span>(Y) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"tr1"</span>, <span class="st">"tr2"</span>)</code></pre></div>
<p>Let’s perform univariate QTL mapping for each of the two traits in the Y matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/qtl2/topics/scan1">scan1</a></span>(<span class="dt">genoprobs =</span> pr, <span class="dt">pheno =</span> Y[ , <span class="dv">1</span>], <span class="dt">kinship =</span> kinship) -&gt;<span class="st"> </span>s1
<span class="kw"><a href="http://www.rdocumentation.org/packages/qtl2/topics/scan1">scan1</a></span>(<span class="dt">genoprobs =</span> pr, <span class="dt">pheno =</span> Y[ , <span class="dv">2</span>], <span class="dt">kinship =</span> kinship) -&gt;<span class="st"> </span>s2</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(s1, pm)</code></pre></div>
<p><img src="testing-close-linkage-vs-pleiotropy_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(s2, pm)</code></pre></div>
<p><img src="testing-close-linkage-vs-pleiotropy_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/qtl2/topics/find_peaks">find_peaks</a></span>(s1, <span class="dt">map =</span> pm)
<span class="co">#&gt;   lodindex lodcolumn chr      pos       lod</span>
<span class="co">#&gt; 1        1    pheno1 `2` 69.64501 23.675561</span>
<span class="co">#&gt; 2        1    pheno1 `3` 80.26172  5.563364</span>
<span class="co">#&gt; 3        1    pheno1   X 12.72372 14.349454</span>
<span class="kw"><a href="http://www.rdocumentation.org/packages/qtl2/topics/find_peaks">find_peaks</a></span>(s2, <span class="dt">map =</span> pm)
<span class="co">#&gt;   lodindex lodcolumn chr       pos       lod</span>
<span class="co">#&gt; 1        1    pheno1 `2`  69.64501 19.446244</span>
<span class="co">#&gt; 2        1    pheno1 `3` 127.28312  3.346147</span>
<span class="co">#&gt; 3        1    pheno1   X  12.72372  4.366299</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">out &lt;-<span class="st"> </span><span class="kw"><a href="../reference/scan_pvl.html">scan_pvl</a></span>(<span class="dt">probs =</span> pp, <span class="dt">pheno =</span> Y, <span class="dt">kinship =</span> kinship<span class="op">$</span><span class="st">`</span><span class="dt">2</span><span class="st">`</span>, <span class="dt">start_snp1 =</span> <span class="dv">38</span>, <span class="dt">n_snp =</span> <span class="dv">25</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(dplyr)
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
out <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/tidy_scan_pvl.html">tidy_scan_pvl</a></span>(pm2) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/add_intercepts.html">add_intercepts</a></span>(<span class="dt">intercepts_univariate =</span> <span class="kw">c</span>(<span class="fl">69.6</span>, <span class="fl">69.6</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/plot_pvl.html">plot_pvl</a></span>(<span class="dt">phenames =</span> <span class="kw">c</span>(<span class="st">"tr1"</span>, <span class="st">"tr2"</span>))
<span class="co">#&gt; Warning: Removed 48 rows containing missing values (geom_path).</span></code></pre></div>
<p><img src="testing-close-linkage-vs-pleiotropy_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#calc_lrt(out)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#setting-for-close-linkage-v-pleiotropy-likelihood-ratio-test">Setting for close linkage v pleiotropy likelihood ratio test</a></li>
      <li><a href="#installing-qtl2pleio">Installing <code>qtl2pleio</code></a></li>
      <li><a href="#reading-data-from-qtl2data-r-package">Reading data from <code>qtl2data</code> R package</a></li>
      <li><a href="#kinship-calculations">Kinship calculations</a></li>
      <li><a href="#simulating-phenotypes-with-qtl2pleiosim1">Simulating phenotypes with <code><a href="--/reference/sim1-html">qtl2pleio::sim1</a></code></a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Frederick Boehm.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
