<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform model fitting for all ordered pairs of markers in a genomic region of interest — scan_pvl • qtl2pleio</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Perform model fitting for all ordered pairs of markers in a genomic region of interest — scan_pvl" />

<meta property="og:description" content="Perform model fitting for all ordered pairs of markers in a genomic region of interest" />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qtl2pleio</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/testing-close-linkage-vs-pleiotropy.html">Testing close linkage v. pleiotropy</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform model fitting for all ordered pairs of markers in a genomic region of interest</h1>
    
    <p class="hidden name"><code>scan_pvl.Rd</code></p>
    </div>

    
    <p>Perform model fitting for all ordered pairs of markers in a genomic region of interest</p>
    

    <pre class="usage"><span class='fu'>scan_pvl</span>(<span class='no'>probs</span>, <span class='no'>pheno</span>, <span class='no'>kinship</span>, <span class='kw'>covariates</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>start_snp1</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='no'>n_snp</span>,
  <span class='kw'>max_iter</span> <span class='kw'>=</span> <span class='fl'>1e+06</span>, <span class='kw'>max_prec</span> <span class='kw'>=</span> <span class='fl'>1</span>/<span class='fl'>1e+08</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>probs</th>
      <td><p>an array of genotype probabilities for a single chromosome (before eigenrotation)</p></td>
    </tr>
    <tr>
      <th>pheno</th>
      <td><p>a matrix of phenotypes (before eigenrotation)</p></td>
    </tr>
    <tr>
      <th>kinship</th>
      <td><p>a kinship matrix for one chromosome</p></td>
    </tr>
    <tr>
      <th>covariates</th>
      <td><p>a matrix, n subjects by n.cov covariates, where each column is one covariate</p></td>
    </tr>
    <tr>
      <th>start_snp1</th>
      <td><p>index of where to start the scan within probs</p></td>
    </tr>
    <tr>
      <th>n_snp</th>
      <td><p>the number of (consecutive) snps to include in the scan</p></td>
    </tr>
    <tr>
      <th>max_iter</th>
      <td><p>maximum number of iterations for EM algorithm</p></td>
    </tr>
    <tr>
      <th>max_prec</th>
      <td><p>stepwise precision for EM algorithm. EM stops once incremental difference in log likelihood is less than max_prec</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a tibble with d + 1 columns. First d columns indicate the genetic data (by listing the marker ids) used in the design matrix; last is log likelihood</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## define probs</span>
<span class='no'>probs_pre</span> <span class='kw'>&lt;-</span> <span class='fu'>rbinom</span>(<span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>100</span> * <span class='fl'>10</span>, <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>prob</span> <span class='kw'>=</span> <span class='fl'>1</span> / <span class='fl'>2</span>)
<span class='no'>probs</span> <span class='kw'>&lt;-</span> <span class='fu'>array</span>(<span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>probs_pre</span>, <span class='kw'>dim</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>100</span>, <span class='fl'>1</span>, <span class='fl'>10</span>))
<span class='fu'>dimnames</span>(<span class='no'>probs</span>)<span class='kw'>[[</span><span class='fl'>3</span>]] <span class='kw'>&lt;-</span> <span class='fu'>paste0</span>(<span class='st'>"Marker"</span>, <span class='fl'>1</span>:<span class='fl'>10</span>)
<span class='co'># define Y</span>
<span class='no'>Y_pre</span> <span class='kw'>&lt;-</span> <span class='fu'>runif</span>(<span class='fl'>200</span>)
<span class='no'>Y</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>Y_pre</span>, <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='fl'>100</span>)
<span class='no'>covariates</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>c</span>(<span class='fu'>runif</span>(<span class='fl'>99</span>), <span class='fl'>NA</span>), <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>)
<span class='no'>cov2</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>c</span>(<span class='no'>covariates</span>[<span class='fl'>1</span>:<span class='fl'>99</span>], <span class='fl'>10</span>), <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>)
<span class='no'>Y2</span> <span class='kw'>&lt;-</span> <span class='no'>Y</span>
<span class='no'>Y2</span>[<span class='fl'>1</span>, <span class='fl'>2</span>] <span class='kw'>&lt;-</span> <span class='fl'>NA</span>
<span class='fu'>scan_pvl</span>(<span class='kw'>probs</span> <span class='kw'>=</span> <span class='no'>probs</span>, <span class='kw'>pheno</span> <span class='kw'>=</span> <span class='no'>Y</span>, <span class='kw'>kinship</span> <span class='kw'>=</span> <span class='fu'>diag</span>(<span class='fl'>100</span>), <span class='kw'>start_snp1</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>n_snp</span> <span class='kw'>=</span> <span class='fl'>10</span>)</div><div class='output co'>#&gt; # A tibble: 100 x 3
#&gt;    Var1     Var2    loglik
#&gt;    &lt;chr&gt;    &lt;chr&gt;    &lt;dbl&gt;
#&gt;  1 Marker1  Marker1  -162.
#&gt;  2 Marker2  Marker1  -173.
#&gt;  3 Marker3  Marker1  -192.
#&gt;  4 Marker4  Marker1  -170.
#&gt;  5 Marker5  Marker1  -198.
#&gt;  6 Marker6  Marker1  -172.
#&gt;  7 Marker7  Marker1  -184.
#&gt;  8 Marker8  Marker1  -177.
#&gt;  9 Marker9  Marker1  -172.
#&gt; 10 Marker10 Marker1  -165.
#&gt; # ... with 90 more rows</div><div class='input'><span class='fu'>scan_pvl</span>(<span class='kw'>probs</span> <span class='kw'>=</span> <span class='no'>probs</span>, <span class='kw'>pheno</span> <span class='kw'>=</span> <span class='no'>Y2</span>, <span class='kw'>kinship</span> <span class='kw'>=</span> <span class='fu'>diag</span>(<span class='fl'>100</span>), <span class='kw'>start_snp1</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>n_snp</span> <span class='kw'>=</span> <span class='fl'>10</span>)</div><div class='output co'>#&gt; <span class='message'>1 subjects dropped due to missing values</span></div><div class='output co'>#&gt; # A tibble: 100 x 3
#&gt;    Var1     Var2    loglik
#&gt;    &lt;chr&gt;    &lt;chr&gt;    &lt;dbl&gt;
#&gt;  1 Marker1  Marker1  -156.
#&gt;  2 Marker2  Marker1  -170.
#&gt;  3 Marker3  Marker1  -186.
#&gt;  4 Marker4  Marker1  -166.
#&gt;  5 Marker5  Marker1  -191.
#&gt;  6 Marker6  Marker1  -169.
#&gt;  7 Marker7  Marker1  -178.
#&gt;  8 Marker8  Marker1  -174.
#&gt;  9 Marker9  Marker1  -165.
#&gt; 10 Marker10 Marker1  -162.
#&gt; # ... with 90 more rows</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Frederick Boehm.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

