<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform model fitting for all ordered pairs of markers in a genomic region of interest â€” scan_pvl â€¢ qtl2pleio</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Perform model fitting for all ordered pairs of markers in a genomic region of interest â€” scan_pvl" />

<meta property="og:description" content="`scan_pvl` calculates log likelihood for d-variate phenotype model fits. Inputted parameter `start_snp` indicates where in the `probs` object to start the scan." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qtl2pleio</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/testing-pleiotropy-vs-separate-qtl.html">Testing pleiotropy v separate QTL in multiparental populations</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/fboehm/qtl2pleio">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform model fitting for all ordered pairs of markers in a genomic region of interest</h1>
    <small class="dont-index">Source: <a href='https://github.com/fboehm/qtl2pleio/blob/master/R/scan_pvl.R'><code>R/scan_pvl.R</code></a></small>
    <div class="hidden name"><code>scan_pvl.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>`scan_pvl` calculates log likelihood for d-variate phenotype model fits. Inputted parameter `start_snp` indicates where in the `probs` object to start the scan.</p>
    
    </div>

    <pre class="usage"><span class='fu'>scan_pvl</span>(<span class='no'>probs</span>, <span class='no'>pheno</span>, <span class='kw'>kinship</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>addcovar</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>start_snp</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='no'>n_snp</span>, <span class='kw'>max_iter</span> <span class='kw'>=</span> <span class='fl'>10000</span>, <span class='kw'>max_prec</span> <span class='kw'>=</span> <span class='fl'>1</span>/<span class='fl'>1e+08</span>,
  <span class='kw'>n_cores</span> <span class='kw'>=</span> <span class='fl'>1</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>probs</th>
      <td><p>an array of founder allele probabilities for a single chromosome</p></td>
    </tr>
    <tr>
      <th>pheno</th>
      <td><p>a matrix of phenotypes</p></td>
    </tr>
    <tr>
      <th>kinship</th>
      <td><p>a kinship matrix for one chromosome</p></td>
    </tr>
    <tr>
      <th>addcovar</th>
      <td><p>a matrix, n subjects by n.cov, of additive covariates, where each column is one numeric additive covariate.</p></td>
    </tr>
    <tr>
      <th>start_snp</th>
      <td><p>index of where to start the scan within probs</p></td>
    </tr>
    <tr>
      <th>n_snp</th>
      <td><p>the number of (consecutive) markers to include in the scan</p></td>
    </tr>
    <tr>
      <th>max_iter</th>
      <td><p>maximum number of iterations for EM algorithm</p></td>
    </tr>
    <tr>
      <th>max_prec</th>
      <td><p>stepwise precision for EM algorithm. EM stops once incremental difference in log likelihood is less than max_prec</p></td>
    </tr>
    <tr>
      <th>n_cores</th>
      <td><p>number of cores to use for calculations</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a tibble with d + 1 columns. First d columns indicate the genetic data (by listing the marker ids) used in the design matrix; last is log likelihood</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The function first discards individuals with one or more missing phenotypes or missing covariates.
It then infers variance components, Vg and Ve. Both Vg and Ve
are d by d covariance matrices. It uses an expectation maximization algorithm, as
implemented in the `gemma2` R package. `gemma2` R package is an R implementation of the
GEMMA algorithm (Zhou &amp; Stephens 2014 Nature methods).
Note that variance components are fitted on a model that uses the d-variate phenotype
but contains no genetic information. This model does, however,
use the specified covariates (after dropping dependent columns
in the covariates matrix).
These inferred covariance matrices, \(\hat{Vg}\) and \(\hat{Ve}\),
are then used in subsequent model fitting via
generalized least squares.
Generalized least squares model fitting is applied to every d-tuple of
markers within the specified genomic region for `scan_pvl`.
A progress bar appears *after* inferring Vg and Ve and records progress through the
d-tuples of markers that constitute the d-variate scan.
For a single d-tuple of markers, we fit the model:
$$vec(Y) = Xvec(B) + vec(G) + vec(E)$$ where
$$G \sim MN(0, K, \hat{Vg})$$ and $$E \sim MN(0, I, \hat{Ve})$$ where \(MN\) denotes the matrix-variate
normal distribution with three parameters: mean matrix, covariance among rows, and
covariance among columns. \(vec\) denotes the vectorization operation, ie, stacking by columns.
\(K\) is a kinship matrix, typically calculated by leave-one-chromosome-out methods.
\(Y\) is the n by d phenotypes matrix. \(X\) is a block-diagonal nd by fd matrix consisting of
d blocks each of dimension n by f. Each n by f block (on the diagonal) contains a matrix of
founder allele probabilities for the n subjects at a single marker. The off-diagonal blocks
have only zero entries.
The log-likelihood is returned for each model. The outputted object is a tibble with
d + 1 columns. The first d columns specify the markers used in the corresponding model fit, while
the last column specifies the log-likelihood value at that d-tuple of markers.
Note also that the covariates matrix must be a set of linearly independent columns.
scan_pvl checks for linear independence of columns in covariates and returns an error
if the columns form a linear dependent set of vectors.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Knott SA, Haley CS (2000) Multitrait
least squares for quantitative trait loci detection.
Genetics 156: 899â€“911.</p>
<p>Jiang C, Zeng ZB (1995) Multiple trait analysis
of genetic mapping for quantitative trait loci.
Genetics 140: 1111-1127.</p>
<p>Zhou X, Stephens M (2014) Efficient multivariate linear
mixed model algorithms for genome-wide association studies.
Nature methods 11:407-409.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># read data</span>
<span class='no'>iron</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>qtl2</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/qtl2/topics/read_cross2'>read_cross2</a></span>(<span class='fu'>system.file</span>(<span class='st'>"extdata"</span>, <span class='st'>"iron.zip"</span>, <span class='kw'>package</span><span class='kw'>=</span><span class='st'>"qtl2"</span>))
<span class='co'># insert pseudomarkers into map</span>
<span class='no'>map</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>qtl2</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/qtl2/topics/insert_pseudomarkers'>insert_pseudomarkers</a></span>(<span class='no'>iron</span>$<span class='no'>gmap</span>, <span class='kw'>step</span><span class='kw'>=</span><span class='fl'>1</span>)
<span class='co'># calculate genotype probabilities</span>
<span class='no'>probs</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>qtl2</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/qtl2/topics/calc_genoprob'>calc_genoprob</a></span>(<span class='no'>iron</span>, <span class='no'>map</span>, <span class='kw'>error_prob</span><span class='kw'>=</span><span class='fl'>0.002</span>)
<span class='co'># grab phenotypes and covariates; ensure that covariates have names attribute</span>
<span class='no'>pheno</span> <span class='kw'>&lt;-</span> <span class='no'>iron</span>$<span class='no'>pheno</span>
<span class='co'># leave-one-chromosome-out kinship matrices</span>
<span class='no'>kinship</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>qtl2</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/qtl2/topics/calc_kinship'>calc_kinship</a></span>(<span class='no'>probs</span>, <span class='st'>"loco"</span>)$<span class='no'>`1`</span>
<span class='co'># get founder allele probabilites</span>
<span class='no'>probs</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>qtl2</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/qtl2/topics/genoprob_to_alleleprob'>genoprob_to_alleleprob</a></span>(<span class='no'>probs</span>)$<span class='no'>`1`</span>
<span class='no'>ac</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>as.numeric</span>(<span class='no'>iron</span>$<span class='no'>covar</span>$<span class='no'>sex</span> <span class='kw'>==</span> <span class='st'>"m"</span>, <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>))
<span class='fu'>colnames</span>(<span class='no'>ac</span>) <span class='kw'>&lt;-</span> <span class='st'>"sex"</span>
<span class='fu'>rownames</span>(<span class='no'>ac</span>) <span class='kw'>&lt;-</span> <span class='fu'>rownames</span>(<span class='no'>probs</span>)
<span class='fu'>scan_pvl</span>(<span class='kw'>probs</span> <span class='kw'>=</span> <span class='no'>probs</span>, <span class='kw'>pheno</span> <span class='kw'>=</span> <span class='no'>pheno</span>, <span class='kw'>kinship</span> <span class='kw'>=</span> <span class='no'>kinship</span>, <span class='kw'>addcovar</span> <span class='kw'>=</span> <span class='no'>ac</span>,
<span class='kw'>start_snp</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>n_snp</span> <span class='kw'>=</span> <span class='fl'>80</span>, <span class='kw'>n_cores</span> <span class='kw'>=</span> <span class='fl'>1</span>)</div><div class='output co'>#&gt; <span class='message'>starting covariance matrices estimation with data from 284 subjects.</span></div><div class='output co'>#&gt; <span class='message'>covariance matrices estimation completed.</span></div><div class='output co'>#&gt; [90m# A tibble: 6,400 x 3[39m
#&gt;    Var1     Var2    loglik
#&gt;    [3m[90m&lt;chr&gt;[39m[23m    [3m[90m&lt;chr&gt;[39m[23m    [3m[90m&lt;dbl&gt;[39m[23m
#&gt; [90m 1[39m D1Mit18  D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m 2[39m c1.loc28 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m 3[39m c1.loc29 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m 4[39m c1.loc30 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m 5[39m c1.loc31 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m 6[39m c1.loc32 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m 7[39m c1.loc33 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m 8[39m c1.loc34 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m 9[39m c1.loc35 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m10[39m c1.loc36 D1Mit18 -[31m[4m3[24m23[39m[31m0[39m[31m.[39m
#&gt; [90m# ... with 6,390 more rows[39m</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Frederick Boehm.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

